### ISSUE TITLE
chore: add performance regression test infrastructure.

### ISSUE THREAD

**Issue by monperrus** [INRIA/spoon#4084]
#4072 shows that performance can be seriously impacted by small changes.

Today, with our CI, we would not detect any serious performance regression.

Goal: add some performance regression test infrastructure, in Github Actions, Jenkins or elsewhere if appropriate.

**Comment by slarse** [comment]
Ping @MartinWitt @SirYwell, suggestions?

**Comment by SirYwell** [comment]
*Note that I don't have any experience with performance regression testing.*

I've discussed some aspects with @MartinWitt and @I-Al-Istannen in the last days. I'll try to sum up the ideas and problems we have come across.

## Where to run the tests

To get reproducible and consistent results. Means, e.g. a CI with a vCPU is likely insufficient.

## Features vs bug fixes vs refactoring

Depending on the changes, one need to decide if a regression is acceptable or not. For example, fixing a bug might *require* a certain grade of regression while features shouldn't really affect the performance of existing code. Refactoring should have the least impact in my opinion.

## What should be tested

As there is no magic way to determine the actual performance of the existing code without running it, it's important to make sure that relevant components are covered by benchmarks. It also makes sense to look into performance of spoon in typical scenarios. For my pull request (#4072), I used the CtGenerationTest first as I didn't set up any microbenchmarks (yet). This was sufficient in that case as it makes a lot of calls into the affected code. 

Also, most scenarios don't have a very high coverage, which can be an advantage as it helps to decrease the scope when looking for the roots of performance issues, but is a disadvantage as it's hard to cover all aspects with multiple scenarios.

### How to cover new features

Adding support for a new language feature often means that the new code should be covered too. It might work to create scenarios that support new features automatically, but having to care about the coverage of the regression tests manually is an additional burden on the developer.

### What about focusing on hot code?

A different approach would be use typical scenarios to analyze what parts of the code are hot and which aren't. Subsequently, it would be possible to write microbenchmarks for small code pieces. This could help to have more exact results (although it does not necessarily represent "real" use-cases) in measuring that helps to see where a new performance issue was introduced.

Microbenchmarks could be also made a requirement similar to test cases for bugs. That way, reasoning of a specific implementation is easy to look up and one can keep track of different approaches and how well they perform.

----

The best would probably to combine scenarios with specific microbenchmarks. The more data about regressions, the easier it is to actually rate it properly.

A tool to make the data accessible would be e.g. [velcom](https://github.com/IPDSnelting/velcom) (@I-Al-Istannen can tell more about that).
To profile benchmarks with more insights it might make sense to use the [JDK Flight Recorder](https://openjdk.java.net/jeps/328) and/or [async-profiler](https://github.com/jvm-profiling-tools/async-profiler).
For microbenchmarks, [JMH](https://github.com/openjdk/jmh) is recommended.

**Comment by monperrus** [comment]
à propos
What's Wrong with My Benchmark Results? Studying Bad Practices in JMH Benchmarks (IEEE TSE)
http://ieeexplore.ieee.org/document/8747433

**Comment by slarse** [comment]
@I-Al-Istannen Very cool! So, if I'm not mistaken, VelCom is a service that you need to run perpetually? It's not the kind of thing you can just run in a Jenkins job or something along those lines?

**Comment by I-Al-Istannen** [comment]
## Regarding your question
Correct. VelCom executes a benchmark script (which does all the heavy lifting, you can find the one I used above or [here](https://github.com/I-Al-Istannen/velcom-unibench/blob/master/bench_spoon)), parses its output and displays the output in the UI as a graph, comparisons, significant changes and so on. To do that it needs someplace its web server and runner component can live.

I used it here because I wanted to get an overview whether the benchmark I chose (running my JavadocApi indexer on spoon) would provide some useful insights and having it benchmark spoon automatically overnight and displaying the results in a graph was convenient for that.

There is nothing stopping you from running the benchmark script in CI, it is made to be run standalone with only the path to the git repo under test as its lonely input.

## General benchmarking questions

I am not sure what benchmarks you had in mind, the one I created above just takes a (totally useful!) program using spoon and runs it with each revision. So it is a high-level usage benchmark and its data is relevant for as long as the selected version of it is still compatible with spoon. After that you need to update the real-world program and generate new data for the relevant older and new spoon revisions.

If you have some other "typical" projects in mind that could be used for that I could add them and see what happens. The programs ideally don't do too much work on their own or, if they do, are mostly deterministic. JavadocApi was nice for that as almost all lines of code just hit some poor spoon object.

## General benchmarking problems
I don't suppose you have a random unused bare-metal computer lying around so we'd run benchmarks on shared resources. That likely makes micro-benchmarks quite hard to do, but lacking any I haven't tried it.

**Comment by slarse** [comment]
> However, there is nothing stopping you from running the benchmark script in CI, it is made to be run standalone with only the path to the git repo under test as its lonely input.

Let's aim at something like that.

> I am not sure what benchmarks you had in mind

Anything is better than nothing, and right now we have nothing :). But really, parsing and printing are the two core things to benchmark, I believe.

> I don't suppose you have a random unused bare-metal computer lying around

I might have an old laptop that could be used, but I don't have a great way to hook it up to a public IP address. But I believe that our Jenkins instance might be sufficiently stable, at least we could try that first. We don't necessarily need to have super tight bounds on the performance, we primarily want to be able to guard against major regressions as well as gauge improvements.

**Comment by I-Al-Istannen** [comment]
You can throw the bench script I linked above into CI as-is and it will measure some indication of performance, but ideally we would want to grow that. I know too little of the spoon ecosystem to do that, unless you want me to write some synthetic benchmarks that just parse some code and print it out again unchanged.

If the script were to stay in the format VelCom accepts I'd add the master branch of Spoon to our instance and then that can run in parallel and provide some historical data and overview, decoupled from the spoon project itself. How do you plan to trigger the Jenkins job for PRs or random experiments, as you probably want to know regressions before they are merged?

So what do we do now? Is there anything for me to do or do I just leave this issue as-is and wait until you figured something out internally?

**Comment by slarse** [comment]
So, the number one performance metric is parsing. There are simply very few use cases of Spoon where parsing is not part of the equation. For that, we can simply bench parsing a bunch of projects, for example the ones we currently run smoke tests on in Jenkins.

The second most important performance metric is pretty-printing, because it's a common use case, and it traverses almost the entire model.

> How do you plan to trigger the Jenkins job for PRs or random experiments, as you probably want to know regressions before they are merged?

I'd imagine running a meaningful performance bench is going to take some amount of time, so we might run it on a schedule instead of on every PR. Or perhaps we have one tiny benchmark for PRs, and have a larger one that runs weekly.
